Contribution guide (#315)
scverse/spatialdata-io: docs/contributing.md

- **academic labs** that would like to contribute a reader (or converter) for in-house data generation pipelines
- **general users** that would like to improve the library

The `spatialdata-io` repository inherits the technical stack from `spatialdata`, therefore please read also the [contribution guide from the `spatialdata` repository](https://github.com/scverse/spatialdata/blob/main/docs/contributing.md).

The rest of the document will be specific to `spatialdata-io`. Feedback on this document is highly welcomed: you are encouraged to open a PR to improve this guide.
**A summary checklist is provided at the end of the document.**

## Adding a new reader

To contribute to a reader and make code review efficient we kindly ask for:

- the **specification** for the raw data: a link to a (ideally) _public_ resource describing how the data is organized, and/or for a new version of an already supported technology, a changelog descring in details which changes are introduced;
- the **reader** itself: a Python function, generally supported by helper functions/classes/enums, that reads in the raw data and returns a `SpatialData` object;

> - example **data**: preferably one or more (very) small _public_ datasets, and with a permissive license, to represent the data across the various versions and cover eventual edge cases of the raw data format. In alternative, scripts to easily download _public_ data. Working without public data is also possible but discouraged because it makes development and testing harder;

Reviewer: 👍 B-F eventually started requiring the datasets be at least licensed which is what the Zenodo community is for: https://zenodo.org/communities/bio-formats

Reviewer: *Not* doing so leads to lots of trouble down the road.

Author: Thanks, I took this into account and created a Zenodo community for data submission: https://zenodo.org/communities/scverse-spatialdata-submissions/.

## Adding tests: small data vs real data

An effective way to make a specification easier to understand and to improve the code review process is to provide test data.

### When small public test datasets are available

Our preferred way to test a reader is by using (very) small test datasets (e.g. 100kB-10MB). By being small, we can afford to use the datasets in continuous integration tests that are run via GitHub actions. See here the [GitHub workflow](https://github.com/scverse/spatialdata-io/blob/main/.github/workflows/prepare_test_data.yaml) that we use to download the test datasets, and here an [example of a test](https://github.com/scverse/spatialdata-io/blob/d2fe0bc18349093ad2cafce752590117729baee8/tests/test_xenium.py#L47) using the small datasets.

Notes:

- The small test datasets do not need to represent real data and can be entirely artificial.
- If a format specification contains edge cases, we recommend to employ a dataset that covers the various edge cases, or to use multiple test datasets.

### When only real (large) public data is available

> Even when a small dataset is not available, we encourage to produce one. If this is not possible a public real dataset can be used. A possibility is also to use both: use small test datasets for CI testing, and a real dataset for double-checking real world usability and performance.

Reviewer: ```suggestion
Even when a small dataset is not available, we encourage you to produce one. If this is not possible a public real dataset can be used. A possibility is also to use both: use small test datasets for CI testing, and a real dataset for double-checking real world usability and performance.
```

#### The `download.py` script

The `download.py` script simply downloads the raw data inside the subfolder `data` and, if it is zipped, it unzips it. Example [`download.py`](https://github.com/giovp/spatialdata-sandbox/blob/main/visium_hd_3.1.1_io/download.py) for Visium HD data.

When the data does not follow any specification (for instance for in-house data), it sometimes also performs some manipulation on the data and, in such cases, it stores the result in any convenient way. Example [`download.py`](https://github.com/giovp/spatialdata-sandbox/blob/main/merfish/download.py) for in-house MERFISH data.

#### The `to_zarr.py` script

The `to_zarr.py` script: (1) imports a suitable reader from `spatialdata_io`; (2) reads in the raw data from the folder `data` to obtain a `SpatialData` object; (3) saves the object to disk to a new Zarr store called `data.zarr`; (4) as a consistency check, reads the store back in memory to a new `SpatialData` object.

### The importance of public data and permissive licenses

> There are a series of reasons to prefer public datasets provided with a permissive license (e.g. 10x public datasets often uses the [CC BY 4.0 license](https://creativecommons.org/licenses/by/4.0/)):

Reviewer: ```suggestion
There are a series of reasons to prefer public datasets provided with a permissive license (e.g. 10x public datasets often use the [CC BY 4.0 license](https://creativecommons.org/licenses/by/4.0/)):
```

> - It allow the community to use the `download.py` script to download the data. An example of data accessible to the public but when this is not possible is when the data can be downloaded only after completing a questionnaire. For private data, collaborations and community contributions are difficult.

Reviewer: ```suggestion
- It allows the community to use the `download.py` script to download the data. An example of data accessible to the public but when this is not possible is when the data can be downloaded only after completing a questionnaire. For private data, collaborations and community contributions are difficult.
```

- It allow the community to modify (e.g. subset, to make download faster) and rehost the data. This makes it easier to share results and collaborate effectively.

### When only private data is available

If the data cannot be public yet, please reach out to scverse via [Zulip](https://scverse.zulipchat.com/#narrow/channel/443514-spatialdata-dev) via private message, or via email. Please give us access to a private repository where you can upload a `download.py` and `to_zarr.py` script working as described above. In alternative, you can also send us the data directly instead of the `download.py` script.

**Technical notes**

If the `download.py` and `to_zarr.py` scripts require Python imports for packages that are not available in an environment where `spatialdata` is installed, please provide also a `requirement.txt` file (see for instance a [legacy one here](https://github.com/giovp/spatialdata-sandbox/blob/main/merfish/requirements.txt)), specify the dependencies in the scripts as [inline script metadata (PEP 723)](https://docs.astral.sh/uv/guides/scripts/#declaring-script-dependencies).

## What to test

> We encourage to test the reader function and any helper function.

Reviewer: ```suggestion
We encourage testing the reader function and any helper function.
```
